Lib/pathlib/_abc.py

```diff
@@ -706,7 +705,7 @@ def walk(self, top_down=True, on_error=None, follow_symlinks=False):
             if not top_down:
                 paths.append((path, dirnames, filenames))
             try:
-                with path.scandir() as entries:
+                with path._scandir() as entries:
```

Comment:
This won't compose properly from here (best case: we make every arbitrary `Path` construct `os.DirEntry` objects just for the sake of a speedup that won't exist).

I'm afraid `walk` probably has to be overridden on the file system-only subclass to take advantage of file system-only operations.

You _could_ try calling `os.scandir(self)` directly and handling (I assume) the `TypeError` (for paths that don't implement `__fspath__`), but I think it'll still work out cleaner to have a version that relies on `os.scandir` and a version that only uses `self.iterdir`.

Reply:
> This won't compose properly from here (best case: we make every arbitrary `Path` construct `os.DirEntry` objects just for the sake of a speedup that won't exist).

`PathBase` subclasses are only required to implement `iterdir()`. Users may opt-in to implementing their own `_scandir()` to speed up `glob()`, `walk()` and `copy()`


> I'm afraid walk probably has to be overridden on the file system-only subclass to take advantage of file system-only operations.

It's already overridden in `pathlib.Path`, but only because working with strings internally is faster than generating a bunch of intermediate `Path` objects

Reply:
But unless they use a `DirEntry` object, the `stat` calls won't be faster and the speedup won't be there compared to `iterdir`.

A more logical API here would be one that yields the name and a flag indicating whether it can recurse or not, or an API that returns a single `walk()` tuple.

The only thing gained from copying `scandir` is that filesystem paths can be implemented directly, but if they're not even being implemented directly then there's no gain - we should implement an API that's easily provided by non-filesystem paths and efficient for all cases.